### ISSUE TITLE
Create mapAsync variant that instead of Futures creates "per element" actors (or a pool of them)

### ISSUE THREAD

**Issue by drewhk** [akka/akka#18431]
I.e. it is the streaming version of the "per-request-actor" pattern which is useful if the processing per element is very stateful and needs lots of dynamic conversations.

This is a strictly one-to-one transformation, just like the Future one. Things to consider:
 - limit number of parallel actors
 - actor reuse
 - supervision of the worker actors

**Comment by agolubev** [comment]
Should it be something like: ```def mapAsyncWithActor(aliveActors: Int)(f: Out ⇒ Props): Repr[T]```?
With having one supervisor that watching all children actors (alive number <= aliveActors), when next one dies it processes the next message with creating new Props. Not sure if we should reuse actors as you are saying it's "per-request-actor" pattern.
@drewhk, if all this sounds reasonable I can start this task.

**Comment by matheuslima** [comment]
A detail is missing is the type of one-to-one transformation that will performed by this operationr @agolubev. Upstream will send a element that will send to the actor materialized by the function and the downstream will receive what?

**Comment by agolubev** [comment]
@matheuslima, downstream will receive T message that is sent to supervisor actor (second function attribute). Than materialized actor needs to be terminated: by itself or by supervisor. 
Don't think we need to keep pool of actors.

**Comment by matheuslima** [comment]
I guess the actor must be stopped after process element. Makes no sense maintain the instance because this function is designed to be used with stateful processing actors.

**Comment by matheuslima** [comment]
Btw I prefer `def mapAsyncWithActor(aliveActors: Int)(f: (Out,ActorRef) ⇒ Props): Repr[T]`

**Comment by matheuslima** [comment]
Last question, what is the real advantage in use this function instead a simple mapAsync with ask pattern?

**Comment by ktoso** [comment]
Handling termination of the actor.

**Comment by agolubev** [comment]
So supervisor will terminate actor as soon as receives response from user actor. Correct?

**Comment by ktoso** [comment]
So the problem case we want to avoid (that mapAsync has) is:
`mapAsync(el => act ? el)` however act is dead already, so we'll get the timeout instead of the failure eagerly that "hey, that actor is dead already".

Exact semantics of what we need here I'm not yet sure... we should try first to write up the behaviour we'll want from the operator. 

Should it start the actors? I think so. 
Should it re-start the actors? Plugging in our supervision strategy solves that.
What should it do if 1 actor terminates, yet we had 10 of them. Should it gradually "decrease parallelism" since now only 9 are available, or should it terminate the stream.
What if all actors stop - that should terminate the stream I think – with an error or completion? Error seems better here.

What other questions did I miss? Let's discuss before implementing this one.

**Comment by agolubev** [comment]
Should actor get input message on start? Per @ktoso - No. But this is part of suggested API above.
in case of keeping actors alive need something like:
```def mapAsyncWithActor(aliveActors: Int)(factory: (ActorRef) ⇒ Props[Out]): Repr[T]```
so supervisor will send element to users actor and waiting for result as back message.
Should we use timeout here? or just be ok with accidental/deliberate loosing messages? - Don't like this as this make backpressure compicated.
@ktoso, can you send a link to code example of "Plugging in our supervision strategy solves that."

**Comment by agolubev** [comment]
Should different user actors know about each other? (perhaps they can ask supervisor for children - like standard way)

**Comment by hepin1989** [comment]
I we could take a look at `akka.routing.Group` and `akka.routing.Pool` here.and I don't think we could make use of `Props[T]` ,it lives in Akka-typed.

**Comment by agolubev** [comment]
@ktoso, can you take a look at:  https://github.com/akka/akka/issues/18431#issuecomment-168019196
are we good to go with this ticket?

**Comment by rkuhn** [comment]
I think this should be solved differently: this is an Actor-based problem that should be solved with actors, and mapAsync is then enough to tie the solution into a stream.

**Comment by drewhk** [comment]
The basic issue is that ask requires a Timeout and hence not appropriate for cases where you pull messages out from an actor instead of pushing messages into it.

**Comment by rkuhn** [comment]
I don’t see how this comment fits into this ticket: what has been proposed so far is semantically equivalent but a lot more complicated than what I propose.

**Comment by drewhk** [comment]
That is true. But I don't think you are right that this is an actor-based problem. The issue here is to tie the lifecycle of the worker actors to the lifecycle of the stage that has them, I don't think this is trivial to solve by the existing tools.

**Comment by rkuhn** [comment]
Ah, you mean that the stream should own the pool—I meant that a stream should be able to use an external pool with separate lifecycle. This ties into #18142 as well: the pool supervisor I propose would be hooked into the “improved” mapAsync variant that is actor-specific.
